Patch status:  WIP / In development
License identified in Makefile:
# SPDX-License-Identifier: GPL-2.0
License identified for kernel/gcov/Makefile:
# SPDX-License-Identifier: GPL-2.0
License identified in arch/x86/kernel/paravirt.c:
// SPDX-License-Identifier: GPL-2.0-or-later
--- a/Makefile.orig	2023-09-11 21:30:18.095774260 -0700
+++ b/Makefile	2023-09-11 21:31:54.343771920 -0700
@@ -619,7 +619,46 @@ endif
 # Defaults to vmlinux, but the arch makefile usually adds further targets
 all: vmlinux
 
-CFLAGS_GCOV	:= -fprofile-arcs -ftest-coverage \
+CFLAGS_GCOV	:=
+CFLAGS_GCOV_DEFAULT	:= 1
+
+ifeq ($(GCC_PGO_PHASE),GCC_PDI)
+CFLAGS_GCOV	:= -fprofile-arcs -fno-tree-loop-im
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(GCC_PGO_PHASE),GCC_PGI)
+CFLAGS_GCOV	:= -fprofile-generate -fno-tree-loop-im
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(GCC_PGO_PHASE),GCC_PGI_CFG)
+CFLAGS_GCOV	:= -fprofile-arcs -fno-tree-loop-im
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(GCC_PGO_PHASE),GCC_PDO)
+KCFLAGS	:= -fbranch-probabilities -fprofile-dir=$(GCC_PGO_PROFILE_DIR) -fprofile-correction -Wno-error=missing-profile -Wno-error=coverage-mismatch
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(GCC_PGO_PHASE),GCC_PGO)
+KCFLAGS	:= -fprofile-use -fprofile-dir=$(GCC_PGO_PROFILE_DIR) -fprofile-correction -Wno-error=missing-profile -Wno-error=coverage-mismatch
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(GCC_PGO_PHASE),GCC_PGO_CFG)
+KCFLAGS	:= -fprofile-use -fprofile-dir=$(GCC_PGO_PROFILE_DIR) -fprofile-correction -Wno-error=missing-profile -Wno-error=coverage-mismatch
+CFLAGS_GCOV_DEFAULT	= 0
+endif
+
+ifeq ($(CFLAGS_GCOV_DEFAULT),1)
+CFLAGS_GCOV	:= -fprofile-arcs -ftest-coverage $(call cc-option,-fno-tree-loop-im)
+endif
+
+#$(info "GCC_PGO_PHASE:  ${GCC_PGO_PHASE}")
+#$(info "CFLAGS_GCOV:  ${CFLAGS_GCOV}")
+
+CFLAGS_GCOV	+=  \
-	$(call cc-option,-fno-tree-loop-im) \
 	$(call cc-disable-warning,maybe-uninitialized,)
 export CFLAGS_GCOV
--- a/kernel/gcov/Makefile.orig	2023-08-27 14:49:51.000000000 -0700
+++ b/kernel/gcov/Makefile	2023-09-13 15:27:37.344512036 -0700
@@ -1,4 +1,9 @@
 # SPDX-License-Identifier: GPL-2.0
+
+ifeq ($(GCC_PGO_PHASE),GCC_PGI)
+GCOV_PROFILE := n
+endif
+
 ccflags-y := -DSRCTREE='"$(srctree)"' -DOBJTREE='"$(objtree)"'
 
 obj-y := base.o fs.o
--- a/arch/x86/kernel/paravirt.c.orig	2023-09-13 16:41:46.356396964 -0700
+++ b/arch/x86/kernel/paravirt.c	2023-09-13 16:42:31.456395798 -0700
@@ -48,6 +48,7 @@
  * avoid confusing the entry prologues.
  */
 extern void _paravirt_nop(void);
+//
 asm (".pushsection .entry.text, \"ax\"\n"
      ".global _paravirt_nop\n"
      "_paravirt_nop:\n\t"
@@ -57,11 +58,13 @@ asm (".pushsection .entry.text, \"ax\"\n
      ".popsection");
 
 /* identity function, which can be inlined */
+#pragma GCC optimize "-fno-profile-values"
 u32 notrace _paravirt_ident_32(u32 x)
 {
 	return x;
 }
 
+#pragma GCC optimize "-fno-profile-values"
 u64 notrace _paravirt_ident_64(u64 x)
 {
 	return x;
